File: android/core/src/main/kotlin/com/risingpadel/core/liga/LigaModels.kt
package com.risingpadel.core.liga

import kotlinx.serialization.Serializable

/**
 * Dominio de la Liga Personal, portado 1:1 de la app Expo (`padel`, src/types/domain.ts)
 * y espejo del Swift `LigaModels.swift`.
 *
 * La regla que gobierna este fichero: **la copia de seguridad JSON de la app Expo (y de
 * la app de iOS) tiene que importar aquí sin transformación**, y viceversa. Por eso los
 * nombres de campo son los del TypeScript — incluidos los heredados de Padel Band
 * (nivelBand, bandInicio…), que hoy contienen el nivel que mide el reloj — y por eso
 * todo campo tiene default: un backup viejo sin un campo nuevo no puede fallar.
 */

@Serializable
data class LigaSetMarcador(
    val yo: String = "",
    val rival: String = "",
    val tbYo: String = "",
    val tbRival: String = "",
)

@Serializable
data class LigaGolpeSesion(val nombre: String, val nota: Double)

@Serializable
data class LigaGolpeVolumen(val nombre: String, val cantidad: Int)

@Serializable
data class LigaSaludPartido(
    val duracionMin: Int = 0,
    val pulsoMedio: Int? = null,
    val pulsoMax: Int? = null,
    val calorias: Int? = null,
)

@Serializable
data class LigaPuntoProgreso(val minuto: Double, val nivel: Double)

@Serializable
data class LigaFrecuenciaGolpeo(val intervaloMin: Int, val cuentas: List<Int>)

/** Un partido de la liga. El espejo exacto de `Match` en la app Expo. */
@Serializable
data class LigaMatch(
    /** `Date.now()` de JavaScript: milisegundos de época. */
    val id: Long,
    val fecha: String = "",
    val tipo: String = "amistoso",
    val resultado: String = "derrota",
    val posicion: String = "reves",
    val sets: String = "",
    val marcador: List<LigaSetMarcador>? = null,
    val club: String = "",
    val companero: String = "",
    /** Los rivales del partido, en texto libre ("Juan y Pedro"). Campo nuevo de esta
     * app: un backup viejo no lo trae y la app Expo lo ignora. */
    val rivales: String? = null,
    val nivel: Double? = null,
    val nivelBand: Double? = null,
    val mejorGolpe: String? = null,
    val mejorPunt: Double? = null,
    val peorGolpe: String? = null,
    val peorPunt: Double? = null,
    val golpesSesion: List<LigaGolpeSesion>? = null,
    val objetivos: List<Boolean> = listOf(false, false, false),
    val nota: String = "",
    val bandInicio: Double? = null,
    val bandFin: Double? = null,
    val bandMediaJugador: Double? = null,
    val golpesVolumen: List<LigaGolpeVolumen>? = null,
    val totalGolpes: Int? = null,
    val salud: LigaSaludPartido? = null,
    val bandPuntos: List<LigaPuntoProgreso>? = null,
    val frecuenciaGolpeo: LigaFrecuenciaGolpeo? = null,
) {
    /** La métrica estrella de la liga: 2 de 3 objetivos cumplidos. */
    val bienJugado: Boolean get() = objetivos.count { it } >= 2
}

/** Strings a propósito: en la web-app original son inputs de texto y el backup los guarda así. */
@Serializable
data class LigaPerfil(
    val nivelPlaytomic: String = "",
    val nivelBand: String = "",
    val nivelObjetivo: String = "",
    val fechaInicio: String = "",
)

@Serializable
data class LigaAnalisis(
    val lectura: String = "",
    val patrones: List<String> = emptyList(),
    val plan: List<String> = emptyList(),
    val foco: String = "",
    val objetivos: List<String> = emptyList(),
    val fecha: String = "",
    val nPartidos: Int = 0,
)

/**
 * Una temporada de la liga: un tramo de fechas con un objetivo de partidos.
 *
 * Los partidos no llevan referencia a su temporada: cada uno cae en la suya por su
 * `fecha`. Así el shape de los partidos del backup no cambia y los historiales
 * importados de antes de que existieran temporadas funcionan igual.
 */
@Serializable
data class LigaTemporada(
    val id: Long,
    val nombre: String = "",
    /** yyyy-mm-dd. El rango es [inicio, fin]; fin vacío = temporada en curso. */
    val fechaInicio: String = "",
    val fechaFin: String = "",
    /** Cuántos partidos se quiere jugar esta temporada. Null = sin meta de volumen. */
    val objetivoPartidos: Int? = null,
) {
    val enCurso: Boolean get() = fechaFin.isEmpty()

    /** true si el partido cae en el rango. Strings yyyy-mm-dd ordenan como fechas. */
    fun contiene(match: LigaMatch): Boolean =
        match.fecha >= fechaInicio && (fechaFin.isEmpty() || match.fecha <= fechaFin)
}

/** El estado completo de la liga: lo que guarda el fichero y lo que exporta el backup. */
@Serializable
data class LigaState(
    val matches: List<LigaMatch> = emptyList(),
    val objetivos: List<String> = emptyList(),
    val perfil: LigaPerfil = LigaPerfil(),
    val analisis: LigaAnalisis? = null,
    val analisisHistorial: List<LigaAnalisis> = emptyList(),
    /** Campo nuevo de esta app: un backup viejo no lo trae y la app Expo lo ignora. */
    val temporadas: List<LigaTemporada> = emptyList(),
)

/** Las métricas de temporada que ya usa iOS, para que Android pinte las mismas. */
/**
 * El cara a cara con una persona: cuántas veces has jugado contra (o con) ella y cómo
 * fue. `ultimos` son los resultados más recientes, true = victoria, el último el más
 * nuevo — para pintar la mini-racha VVDVV.
 */
@Serializable
data class LigaCaraACara(
    val nombre: String,
    val partidos: Int,
    val victorias: Int,
    val ultimos: List<Boolean>,
) {
    val pctVictorias: Int get() = if (partidos == 0) 0 else victorias * 100 / partidos
}

object LigaMetrics {

    /**
     * Estadísticas contra cada rival. El campo `rivales` es texto libre ("Juan y
     * Pedro"), así que se separa por comas, "y", "/" o "&" y se agrupa por nombre
     * normalizado (espacios fuera, mayúsculas da igual); se enseña la grafía de la
     * primera vez. Orden: más partidos primero, y a igualdad, alfabético.
     */
    fun caraACara(matches: List<LigaMatch>): List<LigaCaraACara> =
        agrupa(matches) { partirNombres(it.rivales ?: "") }

    /** Lo mismo, pero con la pareja: con quién juegas y cómo os va. */
    fun conPareja(matches: List<LigaMatch>): List<LigaCaraACara> =
        agrupa(matches) { partirNombres(it.companero) }

    /** "Juan y Pedro" → ["Juan", "Pedro"]. Separadores: coma, " y ", "/", "&". */
    fun partirNombres(texto: String): List<String> =
        texto.split(",", " y ", "/", "&")
            .map { it.trim() }
            .filter { it.isNotEmpty() }

    private fun agrupa(
        matches: List<LigaMatch>,
        nombresDe: (LigaMatch) -> List<String>,
    ): List<LigaCaraACara> {
        // clave normalizada → (grafía original, resultados en orden cronológico)
        val grafias = LinkedHashMap<String, String>()
        val resultados = LinkedHashMap<String, MutableList<Boolean>>()
        for (match in cronologico(matches)) {
            for (nombre in nombresDe(match)) {
                val clave = nombre.lowercase()
                grafias.getOrPut(clave) { nombre }
                resultados.getOrPut(clave) { mutableListOf() }
                    .add(match.resultado == "victoria")
            }
        }
        return resultados.entries
            .map { (clave, lista) ->
                LigaCaraACara(
                    nombre = grafias.getValue(clave),
                    partidos = lista.size,
                    victorias = lista.count { it },
                    ultimos = lista.takeLast(5),
                )
            }
            .sortedWith(compareByDescending<LigaCaraACara> { it.partidos }.thenBy { it.nombre })
    }

    fun racha(matches: List<LigaMatch>): Int {
        var racha = 0
        for (match in cronologico(matches).reversed()) {
            if (!match.bienJugado) break
            racha++
        }
        return racha
    }

    fun mejorRacha(matches: List<LigaMatch>): Int {
        var mejor = 0
        var actual = 0
        for (match in cronologico(matches)) {
            actual = if (match.bienJugado) actual + 1 else 0
            if (actual > mejor) mejor = actual
        }
        return mejor
    }

    fun pctVictorias(matches: List<LigaMatch>): Int {
        if (matches.isEmpty()) return 0
        return (matches.count { it.resultado == "victoria" } * 100.0 / matches.size).toInt()
    }

    /** Nivel de la sesión: el campo directo o, si falta, la media de la curva. */
    fun nivelDeSesion(m: LigaMatch): Double? {
        m.nivelBand?.let { return it }
        val puntos = listOfNotNull(m.bandInicio, m.bandFin)
        return if (puntos.isEmpty()) null else puntos.average()
    }

    private fun cronologico(matches: List<LigaMatch>) = matches.sortedBy { it.fecha }
}

File: ios/RisingPadel/Liga/LigaModels.swift
import Foundation
import PadelCore

/// Dominio de la Liga Personal, portado 1:1 de la app Expo (`padel`, src/types/domain.ts).
///
/// La regla que gobierna este fichero: **la copia de seguridad JSON de la app Expo tiene
/// que importar aquí sin transformación**, y viceversa. Por eso los nombres de campo son
/// los del TypeScript (fecha, tipo, marcador…), por eso `Perfil` guarda strings donde
/// parecería que van números (son campos de texto en la web-app original) y por eso todo
/// se decodifica con tolerancia: un backup viejo sin un campo nuevo no puede fallar.

/// Un set anotado a mano: strings porque en el formulario original son campos de texto.
struct LigaSetMarcador: Codable, Equatable {
    var yo = ""
    var rival = ""
    var tbYo = ""
    var tbRival = ""
}

struct LigaGolpeSesion: Codable, Equatable {
    var nombre: String
    var nota: Double
}

struct LigaGolpeVolumen: Codable, Equatable {
    var nombre: String
    var cantidad: Int
}

struct LigaSaludPartido: Codable, Equatable {
    var duracionMin: Int
    var pulsoMedio: Int?
    var pulsoMax: Int?
    var calorias: Int?
}

struct LigaPuntoProgreso: Codable, Equatable {
    var minuto: Double
    var nivel: Double
}

struct LigaFrecuenciaGolpeo: Codable, Equatable {
    var intervaloMin: Int
    var cuentas: [Int]
}

/// Un partido de la liga. El espejo exacto de `Match` en la app Expo.
struct LigaMatch: Codable, Equatable, Identifiable {
    /// `Date.now()` de JavaScript: milisegundos de época.
    var id: Int64
    var fecha: String
    var tipo: String
    var resultado: String
    var posicion: String
    var sets: String
    var marcador: [LigaSetMarcador]?
    var club: String
    var companero: String
    /// Los rivales del partido, en texto libre ("Juan y Pedro"). Campo nuevo de
    /// esta app: un backup viejo no lo trae y la app Expo lo ignora.
    var rivales: String?
    var nivel: Double?
    var nivelBand: Double?
    var mejorGolpe: String?
    var mejorPunt: Double?
    var peorGolpe: String?
    var peorPunt: Double?
    var golpesSesion: [LigaGolpeSesion]?
    var objetivos: [Bool]
    var nota: String
    var bandInicio: Double?
    var bandFin: Double?
    var bandMediaJugador: Double?
    var golpesVolumen: [LigaGolpeVolumen]?
    var totalGolpes: Int?
    var salud: LigaSaludPartido?
    var bandPuntos: [LigaPuntoProgreso]?
    var frecuenciaGolpeo: LigaFrecuenciaGolpeo?

    /// Decodificación tolerante campo a campo: un backup de una versión vieja (o de la
    /// web-app original) trae menos campos, y uno futuro puede traer más.
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decode(Int64.self, forKey: .id)
        fecha = try c.decodeIfPresent(String.self, forKey: .fecha) ?? ""
        tipo = try c.decodeIfPresent(String.self, forKey: .tipo) ?? "amistoso"
        resultado = try c.decodeIfPresent(String.self, forKey: .resultado) ?? "derrota"
        posicion = try c.decodeIfPresent(String.self, forKey: .posicion) ?? "reves"
        sets = try c.decodeIfPresent(String.self, forKey: .sets) ?? ""
        marcador = try c.decodeIfPresent([LigaSetMarcador].self, forKey: .marcador)
        club = try c.decodeIfPresent(String.self, forKey: .club) ?? ""
        companero = try c.decodeIfPresent(String.self, forKey: .companero) ?? ""
        rivales = try c.decodeIfPresent(String.self, forKey: .rivales)
        nivel = try c.decodeIfPresent(Double.self, forKey: .nivel)
        nivelBand = try c.decodeIfPresent(Double.self, forKey: .nivelBand)
        mejorGolpe = try c.decodeIfPresent(String.self, forKey: .mejorGolpe)
        mejorPunt = try c.decodeIfPresent(Double.self, forKey: .mejorPunt)
        peorGolpe = try c.decodeIfPresent(String.self, forKey: .peorGolpe)
        peorPunt = try c.decodeIfPresent(Double.self, forKey: .peorPunt)
        golpesSesion = try c.decodeIfPresent([LigaGolpeSesion].self, forKey: .golpesSesion)
        objetivos = try c.decodeIfPresent([Bool].self, forKey: .objetivos) ?? [false, false, false]
        nota = try c.decodeIfPresent(String.self, forKey: .nota) ?? ""
        bandInicio = try c.decodeIfPresent(Double.self, forKey: .bandInicio)
        bandFin = try c.decodeIfPresent(Double.self, forKey: .bandFin)
        bandMediaJugador = try c.decodeIfPresent(Double.self, forKey: .bandMediaJugador)
        golpesVolumen = try c.decodeIfPresent([LigaGolpeVolumen].self, forKey: .golpesVolumen)
        totalGolpes = try c.decodeIfPresent(Int.self, forKey: .totalGolpes)
        salud = try c.decodeIfPresent(LigaSaludPartido.self, forKey: .salud)
        bandPuntos = try c.decodeIfPresent([LigaPuntoProgreso].self, forKey: .bandPuntos)
        frecuenciaGolpeo = try c.decodeIfPresent(LigaFrecuenciaGolpeo.self, forKey: .frecuenciaGolpeo)
    }

    init(
        id: Int64,
        fecha: String,
        tipo: String,
        resultado: String,
        posicion: String = "reves",
        sets: String = "",
        marcador: [LigaSetMarcador]? = nil,
        club: String = "",
        companero: String = "",
        rivales: String? = nil,
        nivel: Double? = nil,
        nivelBand: Double? = nil,
        mejorGolpe: String? = nil,
        mejorPunt: Double? = nil,
        peorGolpe: String? = nil,
        peorPunt: Double? = nil,
        golpesSesion: [LigaGolpeSesion]? = nil,
        objetivos: [Bool] = [false, false, false],
        nota: String = "",
        bandInicio: Double? = nil,
        bandFin: Double? = nil,
        bandMediaJugador: Double? = nil,
        golpesVolumen: [LigaGolpeVolumen]? = nil,
        totalGolpes: Int? = nil,
        salud: LigaSaludPartido? = nil,
        bandPuntos: [LigaPuntoProgreso]? = nil,
        frecuenciaGolpeo: LigaFrecuenciaGolpeo? = nil
    ) {
        self.id = id
        self.fecha = fecha
        self.tipo = tipo
        self.resultado = resultado
        self.posicion = posicion
        self.sets = sets
        self.marcador = marcador
        self.club = club
        self.companero = companero
        self.rivales = rivales
        self.nivel = nivel
        self.nivelBand = nivelBand
        self.mejorGolpe = mejorGolpe
        self.mejorPunt = mejorPunt
        self.peorGolpe = peorGolpe
        self.peorPunt = peorPunt
        self.golpesSesion = golpesSesion
        self.objetivos = objetivos
        self.nota = nota
        self.bandInicio = bandInicio
        self.bandFin = bandFin
        self.bandMediaJugador = bandMediaJugador
        self.golpesVolumen = golpesVolumen
        self.totalGolpes = totalGolpes
        self.salud = salud
        self.bandPuntos = bandPuntos
        self.frecuenciaGolpeo = frecuenciaGolpeo
    }
}

/// Perfil del jugador. Strings a propósito: en la web-app original son inputs de texto y
/// el backup los guarda así.
struct LigaPerfil: Codable, Equatable {
    var nivelPlaytomic = ""
    var nivelBand = ""
    var nivelObjetivo = ""
    var fechaInicio = ""
}

/// Un análisis generado por la IA de la liga.
struct LigaAnalisis: Codable, Equatable {
    var lectura = ""
    var patrones: [String] = []
    var plan: [String] = []
    var foco = ""
    var objetivos: [String] = []
    var fecha = ""
    var nPartidos = 0

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        lectura = try c.decodeIfPresent(String.self, forKey: .lectura) ?? ""
        patrones = try c.decodeIfPresent([String].self, forKey: .patrones) ?? []
        plan = try c.decodeIfPresent([String].self, forKey: .plan) ?? []
        foco = try c.decodeIfPresent(String.self, forKey: .foco) ?? ""
        objetivos = try c.decodeIfPresent([String].self, forKey: .objetivos) ?? []
        fecha = try c.decodeIfPresent(String.self, forKey: .fecha) ?? ""
        nPartidos = try c.decodeIfPresent(Int.self, forKey: .nPartidos) ?? 0
    }

    // El init(from:) tolerante se come el memberwise: hay que devolverlo a mano para
    // que el entrenador pueda construir un análisis.
    init(
        lectura: String = "",
        patrones: [String] = [],
        plan: [String] = [],
        foco: String = "",
        objetivos: [String] = [],
        fecha: String = "",
        nPartidos: Int = 0
    ) {
        self.lectura = lectura
        self.patrones = patrones
        self.plan = plan
        self.foco = foco
        self.objetivos = objetivos
        self.fecha = fecha
        self.nPartidos = nPartidos
    }
}

/// Una temporada de la liga: un tramo de fechas con un objetivo de partidos.
///
/// Los partidos **no** llevan referencia a su temporada a propósito: la temporada es un
/// rango de fechas y cada partido cae en la suya por su `fecha`. Así el backup no cambia
/// de shape en los partidos (que es lo que la app Expo relee), funciona con historiales
/// importados de antes de que existieran temporadas, y mover un partido de fecha lo
/// recoloca solo.
struct LigaTemporada: Codable, Equatable, Identifiable {
    /// Milisegundos de época de su creación, como el id de los partidos.
    var id: Int64
    var nombre = ""
    /// yyyy-mm-dd. El rango es [inicio, fin]; fin vacío = temporada en curso.
    var fechaInicio = ""
    var fechaFin = ""
    /// Cuántos partidos se quiere jugar esta temporada. Nil = sin meta de volumen.
    var objetivoPartidos: Int?

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decode(Int64.self, forKey: .id)
        nombre = try c.decodeIfPresent(String.self, forKey: .nombre) ?? ""
        fechaInicio = try c.decodeIfPresent(String.self, forKey: .fechaInicio) ?? ""
        fechaFin = try c.decodeIfPresent(String.self, forKey: .fechaFin) ?? ""
        objetivoPartidos = try c.decodeIfPresent(Int.self, forKey: .objetivoPartidos)
    }

    init(id: Int64, nombre: String, fechaInicio: String, fechaFin: String = "", objetivoPartidos: Int? = nil) {
        self.id = id
        self.nombre = nombre
        self.fechaInicio = fechaInicio
        self.fechaFin = fechaFin
        self.objetivoPartidos = objetivoPartidos
    }

    var enCurso: Bool { fechaFin.isEmpty }

    /// true si el partido cae en el rango de esta temporada. Comparación de strings
    /// yyyy-mm-dd, que ordena igual que las fechas.
    func contiene(_ match: LigaMatch) -> Bool {
        match.fecha >= fechaInicio && (fechaFin.isEmpty || match.fecha <= fechaFin)
    }
}

/// El estado completo de la liga: lo que guarda el fichero y lo que exporta el backup.
struct LigaState: Codable, Equatable {
    var matches: [LigaMatch] = []
    var objetivos: [String] = []
    var perfil = LigaPerfil()
    var analisis: LigaAnalisis?
    var analisisHistorial: [LigaAnalisis] = []
    /// En orden de creación. Campo nuevo de esta app: un backup viejo no lo trae (la
    /// liga entera se comporta como una sola temporada) y la app Expo lo ignora.
    var temporadas: [LigaTemporada] = []

    init() {}

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        matches = try c.decodeIfPresent([LigaMatch].self, forKey: .matches) ?? []
        objetivos = try c.decodeIfPresent([String].self, forKey: .objetivos) ?? []
        perfil = try c.decodeIfPresent(LigaPerfil.self, forKey: .perfil) ?? LigaPerfil()
        analisis = try c.decodeIfPresent(LigaAnalisis.self, forKey: .analisis)
        analisisHistorial =
            try c.decodeIfPresent([LigaAnalisis].self, forKey: .analisisHistorial)
            ?? (analisis.map { [$0] } ?? [])
        temporadas = try c.decodeIfPresent([LigaTemporada].self, forKey: .temporadas) ?? []
    }
}
